src/lib/types/Web3FunctionContext.ts

```diff
@@ -1,9 +1,29 @@
 import { BigNumber } from "@ethersproject/bignumber";
 import { Log } from "@ethersproject/providers";
 import { Web3FunctionMultiChainProvider } from "../provider/Web3FunctionMultiChainProvider";
+import { Web3FunctionOperation } from "./Web3FunctionOperation";
 import { Web3FunctionUserArgs } from "./Web3FunctionUserArgs";
 
-export interface Web3FunctionContextData {
+export type Web3FunctionContextData =
+  | Web3FunctionOnRunContextData
+  | Web3FunctionOnFailContextData
+  | Web3FunctionOnSuccessContextData;
+export interface Web3FunctionOnRunContextData
+  extends Web3FunctionContextDataBase {
+  operation: "onRun";
+}
+export interface Web3FunctionOnFailContextData
+  extends Web3FunctionContextDataBase {
+  operation: "onFail";
+  onFailReason: FailReason;
```

Comment:
If fail reason is "ExecutionReverted" I think we should also include the tx hash